cpp/src/arrow/compute/row/row_test.cc

```diff
@@ -0,0 +1,121 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#include <numeric>
+
+#include "arrow/compute/row/encode_internal.h"
+#include "arrow/compute/row/row_internal.h"
+#include "arrow/testing/generator.h"
+#include "arrow/testing/gtest_util.h"
+
+namespace arrow {
+namespace compute {
+
+namespace {
+
+Result<RowTableImpl> MakeRowTableFromColumn(const std::shared_ptr<Array>& column,
+                                            int64_t num_rows, int row_alignment,
+                                            int string_alignment) {
+  DCHECK_GE(column->length(), num_rows);
+  MemoryPool* pool = default_memory_pool();
+
+  std::vector<KeyColumnArray> column_arrays;
+  std::vector<Datum> values{column};
+  ExecBatch batch(std::move(values), num_rows);
+  RETURN_NOT_OK(ColumnArraysFromExecBatch(batch, &column_arrays));
+
+  std::vector<KeyColumnMetadata> column_metadatas;
+  RETURN_NOT_OK(ColumnMetadatasFromExecBatch(batch, &column_metadatas));
+  RowTableMetadata table_metadata;
+  table_metadata.FromColumnMetadataVector(column_metadatas, row_alignment,
+                                          string_alignment);
+
+  RowTableImpl row_table;
+  RETURN_NOT_OK(row_table.Init(pool, table_metadata));
+
+  RowTableEncoder row_encoder;
+  row_encoder.Init(column_metadatas, row_alignment, string_alignment);
+  row_encoder.PrepareEncodeSelected(0, num_rows, column_arrays);
+
+  std::vector<uint16_t> row_ids(num_rows);
+  std::iota(row_ids.begin(), row_ids.end(), 0);
+
+  RETURN_NOT_OK(row_encoder.EncodeSelected(&row_table, static_cast<uint32_t>(num_rows),
+                                           row_ids.data()));
+
+  return std::move(row_table);
+}
+
+}  // namespace
+
+TEST(RowTableMemoryConsumption, Encode) {
+  constexpr int64_t num_rows_max = 8192;
+  constexpr int64_t padding_for_vectors = 64;
+
+  ASSERT_OK_AND_ASSIGN(
+      auto fixed_length_column,
+      ::arrow::gen::Constant(std::make_shared<UInt32Scalar>(0))->Generate(num_rows_max));
+  ASSERT_OK_AND_ASSIGN(auto var_length_column,
+                       ::arrow::gen::Constant(std::make_shared<BinaryScalar>("X"))
+                           ->Generate(num_rows_max));
+
+  for (int64_t num_rows : {1023, 1024, 1025, 4095, 4096, 4097}) {
+    // Fixed length column.
+    {
+      SCOPED_TRACE("encoding fixed length column of " + std::to_string(num_rows) +
+                   " rows");
+      ASSERT_OK_AND_ASSIGN(auto row_table,
+                           MakeRowTableFromColumn(fixed_length_column, num_rows,
+                                                  uint32()->byte_width(), 0));
+      ASSERT_NE(row_table.data(0), NULLPTR);
+      ASSERT_NE(row_table.data(1), NULLPTR);
+      ASSERT_EQ(row_table.data(2), NULLPTR);
+
+      int64_t actual_null_mask_size =
+          num_rows * row_table.metadata().null_masks_bytes_per_row;
+      ASSERT_GT(actual_null_mask_size * 2,
+                row_table.buffer_size(0) - padding_for_vectors);
```

Comment:
Should we also check that the buffer size is large enough? Same for other inequalities below.

Reply:
Good catch, will add them.

Reply:
Done.